funsor/pyro/hmm.py

```diff
@@ -300,3 +300,135 @@ def expand(self, batch_shape, _instance=None):
+                                observation_matrix.shape[:-2],
+                                observation_mvn.batch_shape)
+        assert shape[-1] == hidden_cardinality
+        batch_shape, time_shape = shape[:-2], shape[-2:-1]
+        event_shape = time_shape + (obs_dim,)
+
+        # Normalize.
+        initial_logits = initial_logits - initial_logits.logsumexp(-1, True)
+        transition_logits = transition_logits - transition_logits.logsumexp(-1, True)
+
+        # Convert tensors and distributions to funsors.
+        init = (tensor_to_funsor(initial_logits, ("class",)) +
+                dist_to_funsor(initial_mvn, ("class",))(value="state"))
+        trans = (tensor_to_funsor(transition_logits, ("time", "class", "class(time=1)")) +
+                 matrix_and_mvn_to_funsor(transition_matrix, transition_mvn,
+                                          ("time", "class(time=1)"), "state", "state(time=1)"))
+        obs = matrix_and_mvn_to_funsor(observation_matrix, observation_mvn,
+                                       ("time", "class(time=1)"), "state(time=1)", "value")
+        if "class(time=1)" not in set(trans.inputs).union(obs.inputs):
+            raise ValueError("neither transition nor observation depend on discrete state")
+        dtype = "real"
+
+        # Construct the joint funsor.
+        with interpretation(lazy):
+            # TODO perform math here once sequential_sum_product has been
+            #   implemented as a first-class funsor.
+            funsor_dist = Variable("value", obs.inputs["value"])  # a bogus value
+            # Until funsor_dist is defined, we save factors for hand-computation in .log_prob().
+            self._init = init
+            self._trans = trans
+            self._obs = obs
+
+        super(SwitchingLinearHMM, self).__init__(
+            funsor_dist, batch_shape, event_shape, dtype, validate_args)
+
+    # TODO remove this once self.funsor_dist is defined.
+    def log_prob(self, value):
+        ndims = max(len(self.batch_shape), value.dim() - 2)
+        value = tensor_to_funsor(value, ("time",), 1)
+
+        with interpretation(moment_matching):
+            result = self._trans + self._obs(value=value)
+            result = sequential_sum_product(ops.logaddexp, ops.add, result, "time",
+                                            {"class": "class(time=1)", "state": "state(time=1)"})
+            result = result.reduce(ops.logaddexp, frozenset(["class(time=1)", "state(time=1)"]))
+            result += self._init
+            result = result.reduce(ops.logaddexp, frozenset(["class", "state"]))
+
+        result = funsor_to_tensor(result, ndims=ndims)
+        return result
+
+    # TODO remove this once self.funsor_dist is defined.
+    def _sample_delta(self, sample_shape):
+        raise NotImplementedError("TODO")
+
+    def expand(self, batch_shape, _instance=None):
+        new = self._get_checked_instance(DiscreteHMM, _instance)
```

Comment:
`DiscreteHMM`?

Reply:
Good catch. I've added a test for this, then fixed a bunch of similar bugs.